### ISSUE TITLE
Meson install on Ubuntu 18.04 tries to exec /usr/bin/meson which does not exist

### ISSUE THREAD

**Issue by rgov** [mesonbuild/meson#8815]
Installing Meson with `sudo pip3 install meson` on Ubuntu 18.04 does not result in a runnable `meson` executable. This was discussed in #4890 which has been closed for some time.

```
$ meson
bash: /usr/bin/meson: No such file or directory
```

The proposed solution is to update `PATH` to include the install directory of Meson (`/usr/local/bin` on Ubuntu 18.04) and indeed this seems to work, but I believe it is still a Meson bug, because the shell *is* finding and executing `/usr/local/bin/meson`:

```
$ which meson
/usr/local/bin/meson
```

For some reason the program at this path is assuming that there is a file at `/usr/bin/meson`, which does not exist.

Note that when one types `asdfasdf` at the Bash prompt, it does not say `/usr/bin/asdfasdf: No such file or directory`, it says `asdfasdf: No such file or directory`. Therefore if it were simply that the user's `PATH` did not include the right folder, it would say `meson: No such file or directory` and not `/usr/bin/meson: No such file or directory`. 

**system parameters**
* native build
* Ubuntu 18.04
* what Python version are you using e.g. 3.8.0
* what `meson --version` 0.58.0
* what `ninja --version` 1.8.2

**Comment by eli-schwartz** [comment]
The error message comes from bash, not meson, so there's nothing meson could possibly do to help you here. This is NOT a meson issue.

You could consider it a pip issue, perhaps. Though the pip developers will probably tell you it's not their fault, it's the fault of Debian's pip fork which is modified to act like this.

For context, this is what happens when you type words into a bash shell. bash will:
- check to see if it remembered you typing this in before (see: `hash -t meson`), and use the saved location as its first guess for the command name
- try the $PATH

In order for you to get the error message you got, you must have previously had meson installed in /usr/bin and then uninstalled it, reinstalling it to /usr/local/bin

Furthermore, you must have actually *run* meson in your current terminal shell in order to save that (broken) location.

You can erase this by running: `hash -d meson`
Or, erase *all* remembered lookups via: `hash -r` (or by opening up a new terminal window)

As for why it used to be in /usr/bin but now is in /usr/local/bin, I'm going to take a wild guess that you installed meson 0.45.1 in Ubuntu Bionic via `apt install`, decided that version was insufficient for you, then sudo pip3 installed a newer version. Debian-patched pip then promptly did something completely un-pythonic and against the upstream designs of python and pip, and added modules to the /usr install of python, by copying the module files into /usr/local instead.

(Of course, pip did need to uninstall the version you currently had installed.)

**Comment by eli-schwartz** [comment]
And a note on:

> ```
> $ which meson
> /usr/local/bin/meson
> ```

Okay, it's now really bash time! So. Here is the thing. The bash (and sh) shell is pretty old, and has a lot of collected warts from 40 years ago. It's complicated to use correctly, and a bit scary to most people. Despite this, everyone is very fond of *using* it as an interactive shell... but (very) few people actually really learn how it works.

So, a lot of very common "knowledge" about how to do things in the shell is unfortunately completely wrong. The widely recommended use of `which` is one of these things.

Do not use `which`. Ever. Some instructive reading:
https://unix.stackexchange.com/questions/85249/why-not-use-which-what-to-use-then/85250#85250
http://mywiki.wooledge.org/BashFAQ/081

tl;dr

The `which` command must not ever be used in shell scripts, since if you do, it won't work when distributed on other platforms. It doesn't have a reliable printed output, it doesn't reliably return errors when it can't find commands, and on commercial Unixes it's probably a csh script and will print you internal details of your ~/.cshrc even when you run it from a bash shell.

Furthermore, it's completely useless. It's a *worse* version of bash `type -a` or zsh `whence -va` (in zsh, `type -a` works too, since `type` is an alias of `whence -v`).

Even the very best, most reliable versions of `which`, like the one installed in modern versions of Debian/Ubuntu (leading people to erroneously think it is *generally* reliable, because Debian is common), act like `command -v` which works in all POSIX sh shells, including bash, including commercial Unixes with their csh script "which". So... why not use command -v if you have a stern requirement on POSIX sh?

More pertinently, for debugging problems you do not, not, NOT want to know "what is in the PATH". You want to know "what did my shell find and try to run, and did I get snagged by an alias or something?"

```
$ which ls
/usr/bin/ls

$ type -a ls
ls is aliased to `ls -FN --color=auto'
ls is /usr/bin/ls
```

If you're debugging a problem in your shell, you *always* want the latter, far more helpful answer.

"which" is a cargo cult of bad design and bad recommendation. The computing world would genuinely be a better place if it had never existed (it has never been useful), but we can try to ameliorate the problem by educating people to use proper tools instead. :)

Thank you for coming to my TED talk. :)

**Comment by rgov** [comment]
> As for why it used to be in /usr/bin but now is in /usr/local/bin, I'm going to take a wild guess that you installed meson 0.45.1 in Ubuntu Bionic via `apt install`, decided that version was insufficient for you, then sudo pip3 installed a newer version. Debian-patched pip then promptly did something completely un-pythonic and against the upstream designs of python and pip, and added modules to the /usr install of python, by copying the module files into /usr/local instead.

Good deduction.

Thanks! I did know about `command -v` as being more portable than `which`, but I have never considered the cases in which they might differ.
